File: apps/backend/src/executors/agentExecutor.ts
import type { ExecutorResult } from './restExecutor.js';

export interface AgentExecutorInput {
  baseUrl:        string;
  authType:       string | null;
  resolvedSecret: string | null;
  endpoint:       string;
  params?:        Record<string, unknown>;
  body?:          Record<string, unknown>;
  pollUrlTemplate?: string;
}

const POLL_INTERVAL_MS = 2_000;
const MAX_TOTAL_MS     = 60_000;
const MAX_POLL_ATTEMPTS = 30;

function buildAuthHeaders(authType: string | null, resolvedSecret: string | null): Record<string, string> {
  const h: Record<string, string> = {
    'Content-Type': 'application/json',
    'Accept':       'application/json',
  };
  if (!resolvedSecret) return h;
  switch (authType) {
    case 'bearer':  h['Authorization'] = `Bearer ${resolvedSecret}`; break;
    case 'api_key': h['X-API-Key']     = resolvedSecret;             break;
    case 'basic':   h['Authorization'] = `Basic ${Buffer.from(resolvedSecret).toString('base64')}`; break;
  }
  return h;
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

export async function agentExecutor(input: AgentExecutorInput): Promise<ExecutorResult> {
  const { baseUrl, authType, resolvedSecret, endpoint, params, body } = input;

  const base    = baseUrl.replace(/\/$/, '');
  const path    = endpoint.startsWith('/') ? endpoint : `/${endpoint}`;
  const kickUrl = new URL(`${base}${path}`);

  if (params) {
    for (const [k, v] of Object.entries(params)) {
      kickUrl.searchParams.set(k, String(v));
    }
  }

  const headers = buildAuthHeaders(authType, resolvedSecret);

  // ── 1. Kick off the job ─────────────────────────────────────────────────────
  console.log(`[agentExecutor] kickoff → ${kickUrl.toString()}`);
  let kickoffResponse: Response;
  try {
    kickoffResponse = await fetch(kickUrl.toString(), {
      method:  'POST',
      headers,
      body:    JSON.stringify(body ?? {}),
    });
  } catch (err) {
    return { success: false, error: `Agent kickoff failed (${kickUrl.toString()}): ${(err as Error).message}` };
  }

  if (!kickoffResponse.ok) {
    // Read the response body so the user sees the API's actual complaint
    // (e.g. "url: must be a valid URL") instead of a generic status text.
    let detail = '';
    try {
      const raw = await kickoffResponse.text();
      if (raw) {
        // If it's JSON, pluck the most useful field; otherwise show first 200 chars
        try {
          const parsed = JSON.parse(raw);
          detail = parsed?.detail || parsed?.error || parsed?.message
                || JSON.stringify(parsed).slice(0, 200);
        } catch {
          detail = raw.slice(0, 200);
        }
      }
    } catch { /* ignore body read errors */ }

    return {
      success: false,
      error:   detail
        ? `Agent kickoff returned ${kickoffResponse.status} ${kickoffResponse.statusText} — ${detail}`
        : `Agent kickoff returned ${kickoffResponse.status} ${kickoffResponse.statusText}`,
    };
  }

  let kickoffJson: any;
  try {
    kickoffJson = await kickoffResponse.json();
  } catch {
    return { success: false, error: 'Agent kickoff returned non-JSON response' };
  }

  const jobId: string | undefined = kickoffJson?.jobId ?? kickoffJson?.job_id;
  if (!jobId || typeof jobId !== 'string') {
    return { success: false, error: 'Agent kickoff did not return a jobId' };
  }

  // ── 2. Poll for completion ──────────────────────────────────────────────────
  const startedAt = Date.now();
  const pollUrlFromResponse: string | undefined =
    kickoffJson?.poll_url ?? kickoffJson?.pollUrl ?? kickoffJson?.result_url;

  console.log(`[agentExecutor] kickoff response:`, JSON.stringify(kickoffJson));

  let pollUrl: string;
  if (pollUrlFromResponse) {
    pollUrl = pollUrlFromResponse.startsWith('http')
      ? pollUrlFromResponse
      : `${base}${pollUrlFromResponse.startsWith('/') ? '' : '/'}${pollUrlFromResponse}`;
    console.log(`[agentExecutor] poll_url from response → ${pollUrl}`);
  } else if (input.pollUrlTemplate) {
    const resolvedTemplate = input.pollUrlTemplate.replace(/\{\{jobId\}\}/g, jobId);
    pollUrl = resolvedTemplate.startsWith('http')
      ? resolvedTemplate
      : `${base}${resolvedTemplate.startsWith('/') ? '' : '/'}${resolvedTemplate}`;
    console.log(`[agentExecutor] poll_url from template → ${pollUrl}`);
  } else {
    pollUrl = `${base}/public/result/${encodeURIComponent(jobId)}`;
    console.log(`[agentExecutor] poll_url fallback → ${pollUrl}`);
  }

  for (let attempt = 1; attempt <= MAX_POLL_ATTEMPTS; attempt++) {
    if (Date.now() - startedAt >= MAX_TOTAL_MS) {
      return { success: false, error: 'Agent timed out after 60s' };
    }

    await sleep(POLL_INTERVAL_MS);

    let pollResponse: Response;
    try {
      pollResponse = await fetch(pollUrl, { method: 'GET', headers });
    } catch (err) {
      return { success: false, error: `Agent poll failed: ${(err as Error).message}` };
    }

    if (!pollResponse.ok) {
      let pollDetail = '';
      try {
        const raw = await pollResponse.text();
        if (raw) {
          try {
            const parsed = JSON.parse(raw);
            pollDetail = parsed?.detail || parsed?.error || parsed?.message || JSON.stringify(parsed).slice(0, 200);
          } catch {
            pollDetail = raw.slice(0, 200);
          }
        }
      } catch { /* ignore */ }
      return {
        success: false,
        error:   pollDetail
          ? `Agent poll returned ${pollResponse.status} ${pollResponse.statusText} (url: ${pollUrl}) — ${pollDetail}`
          : `Agent poll returned ${pollResponse.status} ${pollResponse.statusText} (url: ${pollUrl})`,
      };
    }

    let pollJson: any;
    try {
      pollJson = await pollResponse.json();
    } catch {
      return { success: false, error: 'Agent poll returned non-JSON response' };
    }

    const status = String(pollJson?.status ?? '').toLowerCase();

    if (status === 'complete' || status === 'completed' || status === 'success' || status === 'done') {
      // Some APIs (Nexus) return the result inline; others nest it under .result.
      // Default to the whole payload so nothing is dropped.
      return { success: true, data: pollJson.result ?? pollJson };
    }
    if (status === 'error' || status === 'failed') {
      return { success: false, error: pollJson.error ?? pollJson.message ?? 'Agent reported an error' };
    }
    // queued, running, processing, in_progress, etc. → keep polling
  }

  return { success: false, error: 'Agent timed out after 30 poll attempts' };
}

File: apps/backend-py/app/executors/agent_executor.py
import asyncio
import base64
import logging
from typing import Any

import httpx

log = logging.getLogger("agentExec")

POLL_INTERVAL_S = 2.0
MAX_TOTAL_S = 60.0
MAX_POLL_ATTEMPTS = 30


def _build_headers(auth_type: str | None, resolved_secret: str | None) -> dict[str, str]:
    headers = {"Content-Type": "application/json", "Accept": "application/json"}
    if not resolved_secret:
        return headers
    if auth_type == "bearer":
        headers["Authorization"] = f"Bearer {resolved_secret}"
    elif auth_type == "api_key":
        headers["X-API-Key"] = resolved_secret
    elif auth_type == "basic":
        encoded = base64.b64encode(resolved_secret.encode()).decode()
        headers["Authorization"] = f"Basic {encoded}"
    return headers


def _extract_detail(body: Any) -> str:
    if isinstance(body, dict):
        return str(body.get("detail") or body.get("error") or body.get("message") or "")[:200]
    if isinstance(body, str):
        return body[:200]
    return ""


async def agent_executor(
    base_url: str,
    auth_type: str | None,
    resolved_secret: str | None,
    endpoint: str,
    params: dict[str, Any] | None = None,
    body: dict[str, Any] | None = None,
    poll_url_template: str | None = None,
) -> dict[str, Any]:
    base = base_url.rstrip("/")
    path = endpoint if endpoint.startswith("/") else f"/{endpoint}"
    kick_url = f"{base}{path}"

    headers = _build_headers(auth_type, resolved_secret)

    async with httpx.AsyncClient(timeout=30.0) as client:
        # 1. Kick off the job
        log.info("kickoff → %s", kick_url)
        try:
            resp = await client.post(
                kick_url,
                headers=headers,
                params={k: str(v) for k, v in (params or {}).items()},
                json=body or {},
            )
        except Exception as exc:
            return {"success": False, "error": f"Agent kickoff failed ({kick_url}): {exc}"}

        if not resp.is_success:
            try:
                raw = resp.json()
            except Exception:
                raw = resp.text
            detail = _extract_detail(raw)
            msg = (
                f"Agent kickoff returned {resp.status_code} {resp.reason_phrase} — {detail}"
                if detail
                else f"Agent kickoff returned {resp.status_code} {resp.reason_phrase}"
            )
            return {"success": False, "error": msg}

        try:
            kickoff_json: dict = resp.json()
        except Exception:
            return {"success": False, "error": "Agent kickoff returned non-JSON response"}

        job_id: str | None = kickoff_json.get("jobId") or kickoff_json.get("job_id")
        if not isinstance(job_id, str) or not job_id:
            return {"success": False, "error": "Agent kickoff did not return a jobId"}

        # Resolve poll URL
        poll_url_from_resp: str | None = (
            kickoff_json.get("poll_url")
            or kickoff_json.get("pollUrl")
            or kickoff_json.get("result_url")
        )
        if poll_url_from_resp:
            poll_url = (
                poll_url_from_resp
                if poll_url_from_resp.startswith("http")
                else f"{base}/{poll_url_from_resp.lstrip('/')}"
            )
        elif poll_url_template:
            resolved = poll_url_template.replace("{{jobId}}", job_id)
            poll_url = resolved if resolved.startswith("http") else f"{base}/{resolved.lstrip('/')}"
        else:
            poll_url = f"{base}/public/result/{job_id}"
            log.warning("poll_url fallback → %s", poll_url)

        log.info("polling → %s", poll_url)

        # 2. Poll for completion
        import time
        started_at = time.monotonic()
        for attempt in range(1, MAX_POLL_ATTEMPTS + 1):
            if time.monotonic() - started_at >= MAX_TOTAL_S:
                return {"success": False, "error": "Agent timed out after 60s"}

            await asyncio.sleep(POLL_INTERVAL_S)

            try:
                poll_resp = await client.get(poll_url, headers=headers)
            except Exception as exc:
                return {"success": False, "error": f"Agent poll failed: {exc}"}

            if not poll_resp.is_success:
                try:
                    raw = poll_resp.json()
                except Exception:
                    raw = poll_resp.text
                detail = _extract_detail(raw)
                msg = (
                    f"Agent poll returned {poll_resp.status_code} {poll_resp.reason_phrase} (url: {poll_url}) — {detail}"
                    if detail
                    else f"Agent poll returned {poll_resp.status_code} {poll_resp.reason_phrase} (url: {poll_url})"
                )
                return {"success": False, "error": msg}

            try:
                poll_json: dict = poll_resp.json()
            except Exception:
                return {"success": False, "error": "Agent poll returned non-JSON response"}

            status = str(poll_json.get("status") or "").lower()
            if status in ("complete", "completed", "success", "done"):
                return {"success": True, "data": poll_json.get("result") or poll_json}
            if status in ("error", "failed"):
                return {
                    "success": False,
                    "error": poll_json.get("error") or poll_json.get("message") or "Agent reported an error",
                }
            # queued / running / processing → keep polling

    return {"success": False, "error": "Agent timed out after 30 poll attempts"}
